FirebaseCore/Tests/Unit/FIRAppTest.m

```diff
@@ -98,7 +105,20 @@ - (void)tearDown {
   _appClassMock = nil;
   _notificationCenter = nil;
   _mockCoreDiagnosticsConnector = nil;
+  _mockHeartbeatLogger = nil;
+
+  [super tearDown];
+}
 
++ (void)tearDown {
+  // We stop mocking `FIRHeartbeatLogger` in the class `tearDown` method to
+  // prevent interfering with other tests that use the real `FIRHeartbeatLogger`.
+  // Doing this in the instance `tearDown` causes test failures due to a race
+  // condition between `NSNoticationCenter` and `OCMVerifyAllWithDelay`.
+  // Affected tests:
+  // - testCoreDiagnosticsLoggedWhenAppDidBecomeActive
+  // - testHeartbeatLogIsAttemptedWhenAppDidBecomeActive
+  [OCMClassMock([FIRHeartbeatLogger class]) stopMocking];
```

Comment:
I think this should be `[_mockHeartbeatLogger stopMocking]`, no?

Reply:
No, because this is the _class_ `tearDown` method.

Reply:
Hm, I wonder if this is required at all. It _should_ stop mocking when it deallocates on its own, I know we've added it before on its own though. This seems like we're mocking it again then stopping. Have you tried to see if tests succeed as expected without this?

Reply:
Yes, agreed. Without it, this is what happens:
<img width="783" alt="image" src="https://user-images.githubusercontent.com/36927374/162222641-f5444367-038a-4cdb-a73e-7438a9b72fcc.png">

The `FIRHeartbeatLoggerTests` fail because `OCMock` is still mocking the class (I'm not sure why because as you said: should stop mocking when it deallocates on its own).

When I first came across the issue, I tried to move the `stopMocking` call to the instance `tearDown` method. When doing this, OCMock unregisters the class for mocking so the other test files (`FIRHeartbeatLoggerTests`) aren't affected. But then, a race condition emerges between between OCMock and the way we wait for the `appDidBecomeActive` notification in: 
- testCoreDiagnosticsLoggedWhenAppDidBecomeActive
- testHeartbeatLogIsAttemptedWhenAppDidBecomeActive

Fwiw, I tried waiting for the notification in various ways with [XCTNSNotificationExpectation](https://developer.apple.com/documentation/xctest/xctnsnotificationexpectation?language=objc) but I couldnt get it to work.

<img width="875" alt="image" src="https://user-images.githubusercontent.com/36927374/162223298-6ad7fee0-448d-4c3a-8c67-a6030292f903.png">

Reply:
Strange, thanks a bunch for the info! LGTM